test/e2e/e2e_test.go

```diff
@@ -151,6 +151,10 @@ var _ = ginkgo.ReportAfterEach(func(report ginkgo.SpecReport) {
 	progressReporter.ProcessSpecReport(report)
 })
 
+var _ = ginkgo.ReportBeforeSuite(func(report ginkgo.Report) {
```

Comment:
I honestly can't remember the details, and don't have much time to dig into the code, so sorry for the questions:

does this means this reporter always run? or is still optional?
is it enabled or disabled by default in the CI?

Reply:
@aojea the report is only enabled for e2e test binary: https://github.com/kubernetes/kubernetes/blob/2bb77a13b1a709e0208c6ba9ba5323a5e54f79d6/test/e2e/e2e.go#L112

and it's not used anywhere else as far as I know.

the progress reporter can be treated as *optional* if end user doesn't set the `progress-report-url`, pls see:
https://github.com/kubernetes/kubernetes/blob/2bb77a13b1a709e0208c6ba9ba5323a5e54f79d6/test/e2e/reporters/progress.go#L70-L71

and the comments for that flag,
https://github.com/kubernetes/kubernetes/blob/2bb77a13b1a709e0208c6ba9ba5323a5e54f79d6/test/e2e/framework/test_context.go#L351


the conclusion from me is that this reporter cannot be disabled manually, it's enabled by default, but it's only matter if the flag `progress-report-url` is set, otherwise, there will be no msg sent out.

Reply:
yeah, well, it seems I phrased it wrong , but you answer my question

Thanks

Reply:
The reporter used to have the side effect of writing JSON messages to stdout. But I removed that in 1.26.

Now it is as @chendave said, mostly the presence of this reporter can be ignored during normal test runs.

Reply:
@chendave: does this get called on all worker processes or only on one?

I have asked a few times whether sonobuoy uses this report when running tests in parallel, but so far we have only been able to determine that sonobuoy supports parallel testing. It remains unclear to me whether that is getting used.

My suspicion is that this report is going to be broken when running tests in parallel because each process is going to report its own, local process to the remote URL, with no way of consolidating the different (and conflicting) reports.

We can't fix this, but perhaps you can check and document this as a known limitation?

Reply:
@pohly  I tried to use `tcpdump` to dump the data package, 5 testcases are executed / 3 procs / 3 failed and 2 passed. 

here is the data I captured.

```
{"msg":"","total":5,"completed":0,"skipped":0,"failed":0}
{"msg":"Test Suite starting","total":5,"completed":0,"skipped":0,"failed":0}
{"msg":"FAILED [sig-node] NoExecuteTaintManager Single Pod doesn't evict pod with tolerations from tainted nodes","total":0,"completed":0,"skipped":1332,"failed":1,"failures":["[sig-node] NoExecuteTaintManager Single Pod doesn't evict pod with tolerations from tainted nodes"]}
{"msg":"FAILED [sig-node] NoExecuteTaintManager Single Pod evicts pods from tainted nodes","total":5,"completed":0,"skipped":2524,"failed":1,"failures":["[sig-node] NoExecuteTaintManager Single Pod evicts pods from tainted nodes"]}
{"msg":"PASSED [sig-node] NoExecuteTaintManager Single Pod removing taint cancels eviction [Disruptive] [Conformance]","total":0,"completed":1,"skipped":1181,"failed":0}
{"msg":"Test Suite completed","total":0,"completed":1,"skipped":2592,"failed":0}
{"msg":"PASSED [sig-node] NoExecuteTaintManager Single Pod eventually evict pod with finite tolerations from tainted nodes","total":0,"completed":1,"skipped":1399,"failed":1,"failures":["[sig-node] NoExecuteTaintManager Single Pod doesn't evict pod with tolerations from tainted nodes"]}
{"msg":"Test Suite completed","total":0,"completed":1,"skipped":1399,"failed":1,"failures":["[sig-node] NoExecuteTaintManager Single Pod doesn't evict pod with tolerations from tainted nodes"]}
{"msg":"FAILED [sig-node] NoExecuteTaintManager Single Pod pods evicted from tainted nodes have pod disruption condition","total":5,"completed":0,"skipped":3076,"failed":2,"failures":["[sig-node] NoExecuteTaintManager Single Pod evicts pods from tainted nodes","[sig-node] NoExecuteTaintManager Single Pod pods evicted from tainted nodes have pod disruption condition"]}
{"msg":"Test Suite completed","total":5,"completed":0,"skipped":3076,"failed":2,"failures":["[sig-node] NoExecuteTaintManager Single Pod evicts pods from tainted nodes","[sig-node] NoExecuteTaintManager Single Pod pods evicted from tainted nodes have pod disruption condition"]}
```

the reporter before the suite is started is correct.
```
{"msg":"Test Suite starting","total":5,"completed":0,"skipped":0,"failed":0}
```

but yes, each proc will report on it's own and here are three items reported after the suite is finished, `Test Suite completed`.

Reply:
> does this get called on all worker processes or only on one?

for the msg print, there is only one item for `ReportBeforeSuite`.

Reply:
As I suspected, the messages from the additional workers don't have the right `total`. The other stats in "Test Suite completed" are also incomplete (`completed` is the number of completed tests in that worker, not in the suite).

I don't think that this is a regression, because I don't see how this could have worked differently before with Ginkgo v1, but I am not sure. Before we proceed it would be good to get confirmation from someone familiar with sonobuoy that this behavior is okay. Even better would be a unit test in Kubernetes itself. This could be similar to how I am testing output handling (see https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/log_test.go), except that the unit test needs to be in a separate directory similar to the [cleanup test](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/framework/internal/unittests/cleanup) because defining the Ginkgo test suite is global.

@raosaurabh: does this fix the issue?

/cc @BobyMCbobs

Reply:
> completed is the number of completed tests in that worker, not in the suite

take a look at those msg you will find that if the test passed, the `completed` is set correctly and if it is failed it will be set as *0* instead.

> I don't think that this is a regression

+1

> Even better would be a unit test in Kubernetes itself.

unit test will be great, but I think it's not easy to achieve with an unit test, how do we launch a webservice to response the post request within a UT, and do we really need to test a questionable feature provided by ginkgo.
I will take an investigation with regarding to the unittest but I think it's only nice to have as long as this PR fixed that issue https://github.com/kubernetes/kubernetes/issues/114313

Reply:
Let me prepare a unit test with some TODOs - shouldn't take long, I know that code.

Reply:
> completed is the number of completed tests in that worker, not in the suite

seems like it is calculated as this from `Test Suite completed`, total is 5.
proc1:  1 completed (passed)
proc2:  1 completed (passed), 1 failed
proc3:   2 failed 

this matches with the total test that are executed. `total` is not set correctly, only proc3 has the value set but it's for the entire suite.

Reply:
See https://github.com/chendave/kubernetes/pull/2

Reply:
It cannot run Ginkgo tests in parallel, though.

Reply:
for what it's worth - the behavior y'all are noticing isn't a regression.  it was always broken when running in parallel.  options to fix would include:
- having the server do the aggregation
- spinning up a server on the first parallel process and having other processes report to it, then it report to the central server
- open an issue on ginkgo and ask me how i'd support this (i'd probably introduce `CurrentSuiteReport()` which would allow you to get a synchronized view of the entire suite at any point from any running node.

Reply:
@pohly I think we'd better to create a separate pr for the UT bit, that doesn't seems like related with the issue we would like to fix.

I'm going to reset unless you disagree.

Reply:
That's okay.